editor/py/serve.py

```diff
@@ -38,8 +38,37 @@
 # pylint: disable=invalid-name,no-member,inconsistent-return-statements
 app = flask.Flask(__name__, template_folder='../html')
 
-# Root directory for the editor database (persistent storage).
-DB_ROOT = os.environ.get('ORD_EDITOR_DB', 'db')
+# Root directory for the editor filesystem.
+_DB_ROOT = os.getenv('ORD_EDITOR_DB', 'db')
+# Failure message for check_path.
+_CHECK_PATH_MESSAGE = ''
+_CHECK_PATH_CODE = 403
+
+
+def check_path(path, root=None, user=None):
+    """Checks that a path is safe.
+
+    Specifically, we are looking for directory traversal attacks.
+
+    Args:
+        path: Path to check.
+        root: Root path of the editor filesystem. Defaults to _DB_ROOT.
+        user: String user name. Defaults to flask.g.user.
+
+    Returns:
+        The normalized absolutized path.
+
+    Raises:
+        ValueError: If the path is not deemed safe.
+    """
+    if root is None:
+        root = _DB_ROOT
+    if user is None:
+        user = flask.g.user
+    abspath = os.path.abspath(path)
+    if not abspath.startswith(os.path.abspath(os.path.join(root, user))):
+        raise ValueError(f'path is not safe: {path}')
```

Comment:
I don't know what best practices are, but this seems like a time when using something other than a `ValueError` makes sense. `PermissionError`? Or a custom `AccessError`?

Reply:
Good idea; will do.

Reply:
Changed to `PermissionError`.